parcels/particlefile.py

```diff
@@ -142,46 +125,34 @@ def write(self, pset, time, sync=True, deleted_only=False):
            (self.write_ondelete is False or deleted_only is True):
 
             self.lasttime_written = time
-            if self.type is 'array':
-                # Check if largest particle ID is smaller than the last ID in ParticleFile.
-                # Otherwise, new particles have been added and netcdf will fail
-                if pset.size > 0:
-                    if max([p.id for p in pset]) > self.id[-1]:
-                        logger.error("Number of particles appears to increase. Use type='indexed' for ParticleFile")
-                        exit(-1)
-
-                    # Finds the indices (inds) of the particle IDs in the ParticleFile,
-                    # because particles can have been deleted
-                    pids = [p.id for p in pset]
-                    inds = np.in1d(self.id[:], pids, assume_unique=True)
-                    inds = np.arange(len(self.id[:]))[inds]
-
-                    self.time[inds, self.idx] = time
-                    self.lat[inds, self.idx] = np.array([p.lat for p in pset])
-                    self.lon[inds, self.idx] = np.array([p.lon for p in pset])
-                    self.z[inds, self.idx] = np.array([p.depth for p in pset])
-                    for var in self.user_vars:
-                        getattr(self, var)[inds, self.idx] = np.array([getattr(p, var) for p in pset])
-                    if self.idx == 0:
-                        for var in self.user_vars_once:
-                            getattr(self, var)[inds] = np.array([getattr(p, var) for p in pset])
-                else:
-                    logger.warning("ParticleSet is empty on writing as array")
-
-                self.idx += 1
-            elif self.type is 'indexed':
-                if self.user_vars_once:
-                    logger.warning("Option to_write='once' is not fully functional in indexed mode! Particle properties of such variables are not written for newly added particles.")
-                ind = np.arange(pset.size) + self.idx
-                self.id[ind] = np.array([p.id for p in pset])
-                self.time[ind] = np.array([p.time for p in pset])
-                self.lat[ind] = np.array([p.lat for p in pset])
-                self.lon[ind] = np.array([p.lon for p in pset])
-                self.z[ind] = np.array([p.depth for p in pset])
+            if pset.size > 0:
+                fids = self.id[:]
+                if fids.shape[1] > 0:
+                    fids = fids[:, -1]
+                pids = [p.id for p in pset]
+                in_set = np.in1d(fids, pids, assume_unique=True)
+                oldinds = np.arange(len(fids))[in_set]
+
+                not_in_set = np.in1d(pids, fids, assume_unique=True, invert=True)
```

Comment:
There are two calls to `np.in1d` here, which I fear can be really slow once we get to millions of particles. The issue is that the `particle.id` does not need to be the `trajectory` index in the file (e.g. when multiple particlesets are defined). 
While the current implementation always works, it can be slow. Alternatives could be
1) Get rid of the global `lastID` in `particle.py`, so that the `particle.id` is always the trajectory ID in the file
2) Add an extra Variable `particle.trajinfile` or similar that stores a particle's trajectory ID in the ParticleFile

Option 2 obviously will be easy to implement, but gives extra memory overhead. So I'd first try if I can implement option 1. To be honest, I can't entirely remember anymore exactly *why* we needed the `lastID` variable in `particle.py`